### DISCUSSION TITLE
Use git submodule instead of running separated scripts to add third-party dependencies.

### DISCUSSION THREAD

**Discussion by JackyWoo** [facebookincubator/velox#11203] [Ideas]
Greatings everyone

I am from ClickHouse community and I am new from velox. When I built velox I found it is not very simplified. ClickHouse manages third-party dependencies by git submodule, maybe it is more convenient than separated scripts which velox uses.

1. Build the project and all dependencies in one go to avoid inconsistent compilation options.
2. Convenient to manage the code and build options of third-party dependencies.
3. Try to avoid using dynamic libraries to simplify the deployment environment.

It might be an immature idea., please feel free to express your opinions.

**Comment by JackyWoo** [comment]
@assignUser @majetideepak  could you please take a look at the question?

**Comment by assignUser** [comment]
Hey,
Thanks for you suggestion. We have purposely moved away from git sub-modules as they clutter the repo and cause issues for downstream users like the fact that you can't just download the repo via a github archive link (-> submodules are missing).

We have replaced the submodules with a custom dependency management system based on `FetchContent` which in part also uses git to handle the dependencies but integrates better with cmake. See https://github.com/facebookincubator/velox/tree/main/CMake/resolve_dependency_modules

Some easily installed build tools and deps (flex, bison, openssl, lzo etc.) are not added on purpose, some dependencies are still missing and require the use of the setup-scripts (or other ways to install them). Feel free to contribute to that part of the code if you would like to reduce our reliance on the setup scripts.

**Reply by JackyWoo** [reply]
Thank you!
